### ISSUE TITLE
vue/no-mutating-props "reference-only" option

### ISSUE THREAD

**Issue by cdrini** [vuejs/eslint-plugin-vue#1371]
**What rule do you want to change?**
vue/no-mutating-props

**Does this change cause the rule to produce more or fewer warnings?**
The new "reference-only" option would cause it produce fewer warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option: "reference-only" (i.e. Only error when mutating prop references)

**Please provide some example code that this change will affect:**

```vue
<template>
  <div>
    <input type="range" v-model.number="point.x" >
    <input type="range" v-model.number="point.y" >
  </div>
</template>
<script>
  export default {
    props: {
      point: {
        type: Object,
        required: true
      }
    },
  }
</script>
```

**What does the rule currently do for this code?**
Trigger an error

**What will the rule do after it's changed?**
When the default option is set, the error will trigger. When the option is set to "reference-only", it will not trigger an error.

**Additional context**

(Note there are some issues in the past suggesting this rule be loosened (#1339 , #1314; cc @leosdad @pimlie @decademoon ); here I'm suggesting an option be added instead)

My main reason for wanting an option here, is that modifying the _reference_ of a prop (e.g. `this.myProp = 3`) will cause actual run-time warnings from Vue, cause unexpected behaviour, and is deprecated:

> Due to the new rendering mechanism, whenever the parent component re-renders, the child component’s local changes will be overwritten. - https://vuejs.org/v2/style-guide/#Implicit-parent-child-communication-use-with-caution

Here's the warning Vue throws:

`Avoid mutating a prop directly since the value will be overwritten whenever the parent component re-renders. Instead, use a data or computed property based on the prop's value. Prop being mutated: "number"`

Modifying the _value_ of a prop (but leaving the reference the same) (e.g. `this.myProp.x = 3`), does not throw runtime warnings, and (I believe) does not cause any unexpected behaviour—it's just a code style decision. As long as the actual reference to the object remains unaltered, Vue should not have any issues. 

Here's a sandbox with these two cases for reference: https://codesandbox.io/s/elegant-lederberg-p1ilc

(Note: This is the same way that `const` behaves in JS, so shouldn't be unexpected to most developers:

```js
const x = 3;
x = 7; // Error!

const y = {};
y.foo = 3; // This is fine; it's the _reference_ that's const, not the value.
y = 20; // Error
```
)

I would love it if I could configure linting to error for reference prop modifications (which I _need_ to fix), but to ignore value prop modifications (which I often prefer using in my code). If an option is added called, say "reference-only", then I would use it instead of disabling the rule entirely.

---

P.S. How would one implement my code snippet above without prop value modification? The nicest way I can think of is something like this, which doesn't look particularly better to me.

```vue
<template>
  <div>
    <input type="range" v-model.number="x" >
    <input type="range" v-model.number="y" >
  </div>
</template>
<script>
  export default {
    props: {
      value: {
        type: Object,
        required: true
      }
    },
    data() {
        return {
            x: this.value.x,
            y: this.value.y,
        };
    },
    watch: {
        // Duplicate; could create a `mounted` method and use `$watch` to watch everything
        x(newVal) { this.$emit('input', { x: newVal, y: this.y } },
        y(newVal) { this.$emit('input', { x: this.x, y: newVal } },
    }
  }
</script>
```
(Similar to https://simonkollross.de/posts/vuejs-using-v-model-with-objects-for-custom-components )

**Comment by cdrini** [comment]
Actually, `vue/no-mutating-props` is largely the vue-equivalent to https://eslint.org/docs/rules/no-param-reassign ; there they have:

- `no-param-reassign: ["error", { "props": false }]` as the default (modifying param props is ok)
- `no-param-reassign: ["error", { "props": true }]` disable modifying param props as well

So a better proposal might be to make:

- `vue/no-mutating-props: ["error", { "propProps": true }]` as the default (modifying property props is **NOT** ok)
- `vue/no-mutating-props: ["error", { "propProps": false }]` modifying property props is ok

In a breaking change, renaming this to `vue/no-prop-reassign` might be nice to make it more eslint-y, but I don't think that's worth stalling this. It would also be nice if `propProps: false` was the default like it is for eslint, but anyways :)

**Comment by cdrini** [comment]
^ Updated title/description accordingly

**Comment by mgibas** [comment]
I strongly agree with this issue - `vue/no-mutating-props` without this practically makes v-model unusable except on root level and you end up with something like this:
```
 :value="value.search"
 @input="$emit('input', { ...value, search: $event.target.value })" />
```

IMO [proposal](https://github.com/vuejs/eslint-plugin-vue/issues/256#issuecomment-410448458) that led to creation of this rule had a big flaw:
This:
```
    <input
      :value="todo.text"
      @input="$emit('input', $event.target.value)"
    >
```
Is not an equivalent of this:
```
    <input v-model="todo.text">
```
In this case how would parent know that `todo.text` should be updated on `input` ? I assume that this is not weird scenario where parent is passing whole object but expects that only single property will be modified.

**Comment by nestorrente** [comment]
I agree with this issue too.

In my case, I have several objects that have some common properties (they implement the same interface), and I want to have a common component that represents a form for updating those properties. I pass the object to that form component, and inside the component I use several `<input type="text" v-model="object.property">`.

I think the suggested `vue/no-mutating-props: ["error", { "propProps": false }]` configuration will work perfect for me.

**Comment by mlilback** [comment]
I also agree this needs to change. The [v3 documentation](https://v3.vuejs.org/guide/migration/v-model.html#_3-x-syntax) says:

> In 3.x v-model on the custom component is an equivalent of passing a modelValue prop and emitting an update:modelValue event

So this rule should not be flagged when the prop is an object reference.

**Comment by Liwoj** [comment]
I also think this should be optional. There are cases where using `v-model` on properties of object passed via a props greatly simplifies the code and if the behavior is properly documented (and component properly named), it should be fine. 

Please, give us the choice!

**Comment by ligne13** [comment]
Any update on this one ?

**Comment by douglasg14b** [comment]
Very much agree with this, the rule has been a thorn in our side to the point that it's straight up disabled. We have a rule that dictates we cannot mutate properties of a prop, but fails to provide examples of how we are supposed to do this. And examples I've found are... cumbersome and unweildly, to say the least

The rule as it currently exists has been one of a few examples some dev use as to why "rules are bad and just get in the way". And while that sentiment may be wrong, given the context, they are right.

**Comment by mgibas** [comment]
> The rule as it currently exists has been one of a few examples some dev use as to why "rules are bad and just get in the way". And while that sentiment may be wrong, given the context, they are right, this rule does just get in the way because it fails to be nuanced.

Well said @douglasg14b 

I made multiple implementation where this rule makes sense - working with inputs or set of inputs treated as single, complex, custom input, makes a lot of sense and brought implementation on another level. 

Moment when you’re working with multi step, complex forms, everything crumbles into pieces and makes your code unreadable thanks to bazylion manually handled events flying left and right, that effectively throws v-model out of the window.

I agree with @douglasg14b - this rule is good but current implementation is lacking a lot of nuances.

**Comment by Tofandel** [comment]
Also aggree with this, we need the choice, while it would really not be ok in a lib, in an app, making abstraction when some blocks of components are reused a lot throughout the code when can just copy paste the code add some props and be done with it, but with this rule it forces us to go through an unnecessary clone of the original object and emits which greatly complexify the abstraction process

**Comment by decademoon** [comment]
This is a good rule to abide by but not 100% of the time. There are legitimate cases where this rule should not apply, and doing so is actually detrimental.

**Comment by decademoon** [comment]
I’ve turned off this rule in my codebase because it’s more hassle than it’s worth.

**Comment by Tofandel** [comment]
I can help speed things up with a PR, I already have the tests and the option schema, no I just have the implementation to do (it does look a bit complex to implement correctly though)

**Comment by FloEdelmann** [comment]
A PR would certainly be appreciated! I like `shallowOnly` most. Another suggestion: `ignoreDeepMutations` (`ignore` would be consistent with other rule options).

**Comment by hl037** [comment]
For the implementation, the rule should only warn about direct modification of the `props` object, and not deeper. As I explained in my previous post, modifying a reactive object passed as props should even be encouraged in a model/delegate paradigm

**Comment by ThomasKientz** [comment]
> When objects and arrays are passed as props, while the child component cannot mutate the prop binding, it **will** be able to mutate the object or array's nested properties.

[From Vue's documentation](https://vuejs.org/guide/components/props.html#mutating-object-array-props)

So it eslint shouldn't throw an error when mutate objects or arrays' nested properties.

**Comment by Tofandel** [comment]
I have been digging through the code to make an option for this eslint rule and trying a lot of things but it's not a straightforward implementation, the code is very hard to read through, there is a lot of cases (template mutation, script mutation, this, no this, https://github.com/vuejs/eslint-plugin-vue/blame/master/lib/rules/no-mutating-props.js to get an idea) to account for and my time is quite limited so it will probably take a while until I can get this PR ready

**Comment by padcom** [comment]
Any news? It'd be really awesome to have it configurable and to actually see the problem when it is really there..

**Comment by FloEdelmann** [comment]
No news yet. If nobody works on this and opens a PR, no progress will be made. Asking for updates won't speed that up. So if anybody feels entitled to implement this option, please go ahead!
